c2/cuda/groth16_cuda.cu

```diff
@@ -525,24 +611,57 @@ RustError::by_value generate_groth16_proofs_c(const Assignment<fr_t> provers[],
 
                 {
                     msm_t<bucket_t, point_t, affine_t, scalar_t> msm{nullptr,
-                        tail_msm_l_bases.size()};
+                        std::max({split_vectors_l.tail_msm_scalars[0].size(),
+                                 split_vectors_a.tail_msm_scalars[0].size(),
+                                 split_vectors_b.tail_msm_scalars[0].size()})};
```

Comment:
This parameter determines how large the bucket storage is. In other words it doesn't **have to** be defined by the maximum, any number would actually work, the question is rather how efficient the msms would be overall. So formally speaking an average is more likely to be a better choice. However! My prime argument is that the population counts are a better choice than sizes here. Indeed, we're adding an option to transfer all the scalars, while assuming that there might be as little as 1/2 of "meaningful" ones, and the scaling the bucket storage according to the amount of those is a better fit:-)

Reply:
> while assuming that there might be as little as 1/2 of "meaningful" ones, and the scaling the bucket storage according to the amount of those is a better fit:-)

Ah, makes sense! But currently there's an assert where the `npoints` supplied to constructor is `>=` the `npoints` supplied to `.invoke()`. Should that assert not be there?

Reply:
Right. `npoints` in msm class needs some rethinking. Let's put it aside for now. [On a vaguely related note. I'm moving slice_t to util/gpu_t.cuh, so there will be a breaking point between sppark and this. I'll do the `npoints` refactoring then. So that there will be just one sharp breaking point.]

Reply:
`npoints` adjustments weren't disruptive as I initially imagined, so I committed an update. So that one can switch to [the average of?] population counts now. Or wait for the breaking point mentioned above:-)

The suggested protocol for handing of the breaking point would be to create a pair of PRs, one for sppark and one here, cross-check them and merge at the same time. For this to happen we need to merge this;-)